_posts/2019-06-25-building-a-mono-repo.markdown

```diff
@@ -0,0 +1,66 @@
+
+
+**_Pro_**: Transparency! For better or worse you'll always know when, where, and why each build has been triggered.
+
+**_Cons_**: You've taken complexity that will exist no matter what and moved it from your repo into Jenkins configuration that no one really understands... except maybe you? Adding a new project results in a slew of Jiras for Jenkins setup that are... wait, why are these assigned me? I thought we agree'd people would self service? Wait, I'm the only one who knows how to set these up? Ug, fine.
+
+Also, I haven't used this style of configuration in some time, and if you want handy dandy Pull Request evaluation, it probably won't work anyway.  Unless you attach every build to your PR's and shortcircuit them to not do anyting if a change isn't required. Which, if you do, congratulations, you've just married this approach with the "build your own build system" approach from above.
+
+#### 3. Words Mean What I Say They Mean
+Redefine your idea of a build *is* so that you can run through everything on every change. Build every one of your projects for every Pull Request.  How do you get there?
+* Pull long running tests in a separate suite that runs "afterwards". Do you need to run them on every Pull Request? Do they even fail for reasons that aren't flakey? Are they even providing value or are you a test horder?
+* Use incremental builds instead of full/clean. Yes incremental builds can let issues slip through, but how often does that actually happen in your repo? And is it worth the time it costs to perform them. You can always follow up your incrementals with scheduled full builds throughout the day to catch issues that an incremental would miss.
+* Don't use Scala.
+* Configure your test harnesses so that any tests that need more than a single digit number of seconds to complete are marked as failures. Delete any repeat offenders. Anyone who objects due to their importance is responsible for making them faster.
+* Look the other way when that one project that doesn't have any dependencies in the monorepo some how ended up in its own repo.
+* Accept that there is a natural trade off between the speed of your build (and there for your releases) and the amount of things that can happen to give you confidence in the correctness of a pending change.
+
+**_Pros_**: Admittedly, this is not an option for some, dare say many situations. But if you can, teams I've been on that prioritize build/test discipline like this move blazingly fast. Every minute you add between someone hitting merge and their changing reaching production is that much more encouragement to do one Pull Request instead of two or three. This creates larger Pull Requests which give reviewers that much more excuse to drag their feet on getting them done. This incentivizes authors to create even less (but larger) Pull Requests. So your feedback loop keeps turning, productivity it starts burning.
```

Comment:
last line - is that actually the Fleetwood Mac line? I thought it was "it keeps turning"

Reply:
i really like this bit, i feel like this could be a blog post on its own. pull request size relative to build time.

Reply:
> Big wheel keep on turning
> Proud Mary keep on burning
> And we're rolling, rolling
> Rolling on the river

Bug god is it near impossible to understand.

https://www.youtube.com/watch?v=XfyEpmQM7bw&feature=youtu.be&t=62